internal/pkg/uploader/upload.go

```diff
@@ -0,0 +1,247 @@
+	}
+	id := uid.String()
+
+	// grab required fields that were checked already in validation step
+	agentID, _ := data.Str("agent_id")
+	actionID, _ := data.Str("action_id")
+	source, _ := data.Str("src")
+	docID := fmt.Sprintf("%s.%s", actionID, agentID)
+
+	info := upload.Info{
+		ID:        id,
+		DocID:     docID,
+		AgentID:   agentID,
+		ActionID:  actionID,
+		ChunkSize: MaxChunkSize,
+		Source:    source,
+		Total:     size,
+		Status:    upload.StatusAwaiting,
+		Start:     time.Now(),
+	}
+	chunkCount := info.Total / info.ChunkSize
+	if info.Total%info.ChunkSize > 0 {
+		chunkCount += 1
+	}
+	info.Count = int(chunkCount)
+
+	/*
+		Enrich document with additional server-side fields
+	*/
+
+	if err := data.Put(info.ChunkSize, "file", "ChunkSize"); err != nil {
+		return upload.Info{}, err
+	}
+	if err := data.Put(info.Status, "file", "Status"); err != nil {
+		return upload.Info{}, err
+	}
+	if err := data.Put(id, "upload_id"); err != nil {
+		return upload.Info{}, err
+	}
+	if err := data.Put(info.Start.UnixMilli(), "upload_start"); err != nil {
+		return upload.Info{}, err
+	}
+
+	/*
+		Write to storage
+	*/
+	doc, err := json.Marshal(data)
+	if err != nil {
+		return upload.Info{}, err
+	}
+	_, err = CreateFileDoc(ctx, u.bulker, doc, source, docID)
+	if err != nil {
+		return upload.Info{}, err
+	}
+
+	return info, nil
+}
+
+func (u *Uploader) Chunk(ctx context.Context, uplID string, chunkNum int, chunkHash string) (upload.Info, ChunkInfo, error) {
+	// find the upload, details, and status associated with the file upload
+	info, err := u.GetUploadInfo(ctx, uplID)
+	if err != nil {
+		return upload.Info{}, ChunkInfo{}, err
+	}
+
+	/*
+		Verify Chunk upload can proceed
+	*/
+
+	if info.Expired(u.timeLimit) {
+		return upload.Info{}, ChunkInfo{}, ErrUploadExpired
+	}
+	if !info.StatusCanUpload() {
+		return upload.Info{}, ChunkInfo{}, ErrUploadStopped
+	}
+	if chunkNum < 0 || chunkNum >= info.Count {
+		return upload.Info{}, ChunkInfo{}, ErrInvalidChunkNum
+	}
+
+	return info, ChunkInfo{
+		Pos:  chunkNum,
+		BID:  info.DocID,
+		Last: chunkNum == info.Count-1,
+		Size: int(info.ChunkSize),
+		SHA2: chunkHash,
+	}, nil
+}
+
+func validateUploadPayload(info JSDict) error {
+
+	required := [][]string{
+		{"file", "name"},
+		{"file", "mime_type"},
+		{"action_id"},
+		{"agent_id"},
+		{"src"},
+	}
+
+	for _, fields := range required {
+		if value, ok := info.Str(fields...); !ok || strings.TrimSpace(value) == "" {
+			return fmt.Errorf("%s is required", strings.Join(fields, "."))
+		}
+	}
+
+	if size, ok := info.Int64("file", "size"); !ok {
+		return errors.New("file.size is required")
+	} else if size <= 0 {
+		return fmt.Errorf("invalid file.size: %d", size)
+	}
+	return nil
+}
+
+// Searches for Upload Metadata document in local memory cache if available
+// otherwise, fetches from elasticsearch and caches for next use
+func (u *Uploader) GetUploadInfo(ctx context.Context, uploadID string) (upload.Info, error) {
```

Comment:
[Suggestion | Go convention]
```suggestion
// GetUploadInfo searches for Upload Metadata document in local memory cache if available
// otherwise, fetches from elasticsearch and caches for next use
func (u *Uploader) GetUploadInfo(ctx context.Context, uploadID string) (upload.Info, error) {
```